### ISSUE TITLE
[C++] Can InputType support nested types in the future?

### ISSUE THREAD

**Issue by Light-City** [apache/arrow#35863]
## 1.our problem

we have a kernel, input is struct array, such as struct<sum, count>, When registering the kernel, in order to distinguish between `struct<decimal128, int64>` and `struct<decimal256. int64>` (means: sum column is decimal128 or decimal256), we need to use TypeMatcher to match one.

for example:

```cpp
Status Visit(const Decimal128Type&) {
    kernel = MakeKernel(std::move(InputType(match::AvgFinalDecimal128(), ValueDescr::ARRAY)),
                        AvgSumInit<GroupedAvgFinalImpl<Decimal128Type>>);
    return Status::OK();
  }

  Status Visit(const Decimal256Type&) {
    kernel = MakeKernel(std::move(InputType(match::AvgFinalDecimal256(), ValueDescr::ARRAY)),
                        AvgSumInit<GroupedAvgFinalImpl<Decimal256Type>>);
    return Status::OK();
  }
```

Take `AvgFinalDecimal128` for example:

```cpp
class AvgFinalDecimal128Matcher : public TypeMatcher {
 public:
  AvgFinalDecimal128Matcher() {}

  bool Matches(const DataType& type) const override {
    if (type.id() != Type::STRUCT) {
      return false;
    }
    return (type.fields()[0])->type()->id() == Type::DECIMAL128;
  }
 //......
}
```

In the current code, we need to write two matchers. In fact, we can simplify it to a matcher (the input is a struct input by the user), but currently TypeMatcher does not support input. The constructor is as follows:

```
InputType(std::shared_ptr<TypeMatcher> type_matcher) {}
```

It does not support passing parameters, so will it be possible to support extended parameters in the future?

### Component(s)

C++

**Comment by westonpace** [comment]
I'm not sure I understand your goal.  Is it to only need a single custom type matcher?  Would it be possible to do:

```
class AvgFinalMatcher : public TypeMatcher {
 public:
  AvgFinalMatcher(Type::type type_id) : type_id_(type_id) {}

  bool Matches(const DataType& type) const override {
    if (type.id() != Type::STRUCT) {
      return false;
    }
    return (type.fields()[0])->type()->id() == type_id_;
  }
private:
  Type::type type_id_;
 //......
}

// ...

Status Visit(const Decimal128Type&) {
    kernel = MakeKernel(std::move(InputType(match::AvgFinal(Type::DECIMAL128), ValueDescr::ARRAY)),
                        AvgSumInit<GroupedAvgFinalImpl<Decimal128Type>>);
    return Status::OK();
  }

  Status Visit(const Decimal256Type&) {
    kernel = MakeKernel(std::move(InputType(match::AvgFinal(Type::DECIMAL256), ValueDescr::ARRAY)),
                        AvgSumInit<GroupedAvgFinalImpl<Decimal256Type>>);
    return Status::OK();
  }
```

**Comment by Light-City** [comment]
> I'm not sure I understand your goal. Is it to only need a single custom type matcher? Would it be possible to do:
> 
> ```
> class AvgFinalMatcher : public TypeMatcher {
>  public:
>   AvgFinalMatcher(Type::type type_id) : type_id_(type_id) {}
> 
>   bool Matches(const DataType& type) const override {
>     if (type.id() != Type::STRUCT) {
>       return false;
>     }
>     return (type.fields()[0])->type()->id() == type_id_;
>   }
> private:
>   Type::type type_id_;
>  //......
> }
> 
> // ...
> 
> Status Visit(const Decimal128Type&) {
>     kernel = MakeKernel(std::move(InputType(match::AvgFinal(Type::DECIMAL128), ValueDescr::ARRAY)),
>                         AvgSumInit<GroupedAvgFinalImpl<Decimal128Type>>);
>     return Status::OK();
>   }
> 
>   Status Visit(const Decimal256Type&) {
>     kernel = MakeKernel(std::move(InputType(match::AvgFinal(Type::DECIMAL256), ValueDescr::ARRAY)),
>                         AvgSumInit<GroupedAvgFinalImpl<Decimal256Type>>);
>     return Status::OK();
>   }
> ```

yes, We don't know the specific internal type, it needs to be determined according to the input, that is, this Matches lacks a parameter(just for inputType)

**Comment by Light-City** [comment]
> AvgFinalMatcher

so, how to pass type_id?

original code,
```
InputType(std::shared_ptr<TypeMatcher> type_matcher) {}
```

now code：

```
InputType(std::shared_ptr<TypeMatcher> type_matcher,  Type::type type_id_) {}
```

This kind of interface is really needed.

**Comment by js8544** [comment]
> > AvgFinalMatcher
> 
> so, how to pass type_id?
> 
> original code,
> 
> ```
> InputType(std::shared_ptr<TypeMatcher> type_matcher) {}
> ```
> 
> now code：
> 
> ```
> InputType(std::shared_ptr<TypeMatcher> type_matcher,  Type::type type_id_) {}
> ```
> 
> This kind of interface is really needed.

I think weston's point is that you can pass the type_id to your custom matcher class that inherits `TypeMatcher`, not necessarily to `InputType`.

**Comment by thisisnic** [comment]
This issue has been marked as stale because it has had no activity in the past 2 years. Please remove the stale label or comment below, or this issue will be closed in 14 days.
